benches/benches/bevy_transform/transform_hierarchy.rs

```diff
@@ -0,0 +1,655 @@
+
+    if enable_update {
+        app
+            .add_plugins(TimePlugin)
+            // Updating transforms *must* be done before `CoreSet::PostUpdate`
+            // or the hierarchy will momentarily be in an invalid state.
+            .add_systems(Update, update);
+    }
+
+    // Finish Plugin setup - identical to what the ScheduleRunnerPlugin runner does
+    // We can't use the ScheduleRunnerPlugin since we run app.update() ourselves,
+    // and app.run() can't be called repeatedly when using RunMode::Once
+
+    // Do any of the plugins we use in the benchmarks require any asynchronous
+    // initialization using task pools?
+    // Currently, this is never the case, but the code is kept here as a reference
+    // in case it becomes necessary in the future.
+    const ASYNC_PLUGIN_INIT: bool = false;
+    if ASYNC_PLUGIN_INIT {
+        while !app.ready() {
+            #[cfg(not(target_arch = "wasm32"))]
+            bevy_tasks::tick_global_task_pools_on_main_thread();
+        }
+    }
+    assert!(app.ready());
+
+    app.finish();
+    app.cleanup();
+
+    // Run setup (what would normally happen in the Startup schedule)
+    setup(&mut app.world, cfg);
+
+    app
+}
+
+fn bench_group<M: criterion::measurement::Measurement>(mut group: BenchmarkGroup<M>, cfg: &Cfg) {
+    group.warm_up_time(std::time::Duration::from_secs(2));
+    group.measurement_time(std::time::Duration::from_secs(10));
+
+    // This benchmark tries to measure the cost of the initial transform propagation,
+    // i.e. the first time transform propagation runs after we just added all our entities.
+    group.bench_function("transform_init", |b| {
+        // Building the World (in setup) takes a lot of time, so we shouldn't do that on every
+        // iteration.
+        // Unfortunately, we can't re-use an App directly in iter() the World would no longer be
+        // in its pristine, just initialized state from the second iteration onwards.
+        // Furthermore, it's not possible to clone a pristine World since World doesn't implement
+        // Clone.
+        // As an alternative, we reuse the same App and reset it to a pseudo-pristine state by
+        // simply marking all Parent, Children and Transform components as changed.
+        // This should look like a pristine state to the propagation systems.
+
+        let mut app = build_app(cfg, false);
+
+        app.add_schedule(ResetSchedule, reset_schedule());
+
+        // UnsafeWorldCell is necessary here since we'd otherwise pass a &mut World/&mut App to both
+        // iter_batched closures.
+        let cell = app.world.as_unsafe_world_cell();
+
+        b.iter_batched(
+            || {
+                // reset_schedule.run(unsafe { cell.world_mut() });
+                unsafe { cell.world_mut() }.run_schedule(ResetSchedule);
+
+                cell
+            },
+            |cell| {
+                unsafe { cell.world_mut() }.run_schedule(bevy_app::Main);
```

Comment:
Could you add safety comments here. This requires both closure to never run concurrently. Do you know this for a fact?

Reply:
This should be the case. However, I've thought a little over this and simply replaced the version using iter_batched/UnsafeWorldCell with iter_custom, which doesn't require using UnsafeWorldCell here. Only downside is that it's harder to use custom `Measurement` types with iter_custom, but they aren't currently used in any bevy benchmark.